### PR TITLE
feat: render composite messages UI [part-1]

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

message item does not support composite messages

### Solutions

change the text message composable to accept nullable text and a list of buttons 

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [456](https://10.10.124.134/job/AR-build-pipeline/456/) **failed**.

**Comment by AndroidBob** [comment]
Build [457](https://10.10.124.134/job/AR-build-pipeline/457/) **failed**.

**Comment by AndroidBob** [comment]
Build [458](https://10.10.124.134/job/AR-build-pipeline/458/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.3.0(7064156)(7064156)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.3.0(7064156)(7064156)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt [hunk 1]
```diff
@@ -475,7 +475,31 @@ private fun MessageContent(
...
+                    }
+                    VerticalSpace.x4()
+                }
+                MessageBody(
+                    messageBody = messageContent.messageBody,
+                    isAvailable = !message.isPending && message.isAvailable,
+                    onLongClick = onLongClick,
+                    onOpenProfile = onOpenProfile,
+                    buttonList = messageContent.buttonList,
+                    onButtonClick = { TODO() },
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt]: Just making sure, it will be implemented in a following PR, right?

**Review by alexandreferris — APPROVED** [review]
👌🏻 Just a small question on the `TODO` button click

**Comment by AndroidBob** [comment]
Build [475](https://10.10.124.134/job/AR-build-pipeline/475/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt (hunk 1 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt]: yes

**Review by alexandreferris — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [482](https://10.10.124.134/job/AR-build-pipeline/482/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.3.0(7072433)(7072433)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.3.0(7072433)(7072433)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessageViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import androidx.compose.runtime.mutableStateMapOf
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class CompositeMessageViewModel @Inject constructor() : ViewModel() {
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessageViewModel.kt]:
```suggestion
class CompositeMessagesViewModel @Inject constructor() : ViewModel() {
```

It's not specific to a single message, but rather to all (?) composite messages.

**Review by vitorhugods — APPROVED** [review]
LGTM. Just one suggestion about naming of the VM

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Also tests.

`CompositeMessagesViewModel` has no tests?

**Comment by MohamadJaara** [comment]
> `CompositeMessagesViewModel`

The ViewModel atm does nothing, nevertheless I did add tests with TODO to assert the use case once it is implemented

**Comment by AndroidBob** [comment]
Build [494](https://10.10.124.134/job/AR-build-pipeline/494/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.3.0(7074835)(7074835)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.3.0(7074835)(7074835)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class CompositeMessagesViewModel @Inject constructor() : ViewModel() {
+
+    var pendingButtons = mutableStateMapOf<String, String>()
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt:32]: I am abit confused what this is doing ? From what I read you want to implement some kind of pending state for a button that. then moves to some kind of done state after we do something inside the viewmodel ?

Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt:32]: also this is reaaaaaaaaly focused on the UI related behaviour I think, but since we introduced viewmodel scoped to the Composable component I think we can live with that, but I feel like we are loosin once again what the viewmodel are supposed to be doing 😢

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt [hunk 4]
```diff
@@ -475,7 +481,35 @@ private fun MessageContent(
                     messageBody = messageContent.messageBody,
                     isAvailable = !message.isPending && message.isAvailable,
                     onLongClick = onLongClick,
-                    onOpenProfile = onOpenProfile
+                    onOpenProfile = onOpenProfile,
+                    buttonList = null,
+                    onButtonClick = null,
+                    pendingButton = null,
+                    messageId = message.header.messageId
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt:488]: In my personal opinion it would be create to create a seperate MessageBody component that would be a Composite. It read better when u have a seperate component and see exactly what it takes in as component instead of diving into the function and looking for the if and null checks

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt [hunk 5]
```diff
@@ -444,6 +448,8 @@ private fun MessageContent(
     audioMessagesState: Map<String, AudioState>,
     onAssetClick: Clickable,
     onImageClick: Clickable,
+    onMessageButtonClicked: (messageId: String, buttonId: String) -> Unit,
+    pendingButtonsMap: Map<String, String>,
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt:452]: I wonder how hard this hits the recomposition, every time a button is clicked, everything up to Converastion content will go through recomposition, it will probably skips lots of Composable anyways, but perhaps there is a way to avoid this

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt [hunk 6]
```diff
@@ -475,7 +481,35 @@ private fun MessageContent(
...
+                MessageBody(
+                    messageBody = messageContent.messageBody,
+                    isAvailable = !message.isPending && message.isAvailable,
+                    onLongClick = onLongClick,
+                    onOpenProfile = onOpenProfile,
+                    buttonList = messageContent.buttonList,
+                    onButtonClick = onMessageButtonClicked,
+                    messageId = message.header.messageId,
+                    pendingButton = pendingButtonsMap[message.header.messageId]
                 )
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt]: this all together could be its own Composable, if you remove stuff from MessageBody all this could be done in one place having its own place

**Review by trOnk12 — COMMENTED** [review]
Great job in general ! :) however, I think there is a way to improve it, since Kubaz introduced ViewModel scoped to the Composable, we could instead of having a map and passing it allllll the way through the Composable opening a door for some unnecessary Recomposition, we could create a seperate Composable component and use the ViewModel that you created into that Component, that way each Button will have its own way of manipulating only that one Component.

You can take a look here : ConnectionActionButton for inspiration

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt [hunk 7]
```diff
@@ -90,8 +102,70 @@ internal fun MessageBody(
...
+}
 
-    MarkdownDocument(Parser.builder().extensions(extensions).build().parse(text) as Document, nodeData)
+@Composable
+fun MessageButtonsContent(
+    messageId: String,
+    pendingButton: String?,
+    buttonList: List<MessageButton>,
+    onClick: ((messageId: String, buttonId: String) -> Unit)?
+) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt:124]: I think this would be a good place where the ViewModel could live

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt [hunk 8]
```diff
@@ -90,8 +102,70 @@ internal fun MessageBody(
...
+}
+
+@SuppressLint("RememberReturnType")
+@Composable
+fun MessageButton(
+    messageId: String,
+    button: MessageButton,
+    isPending: Boolean,
+    onClick: ((messageId: String, buttonId: String) -> Unit)?
+) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt:152]: This could be a good place where the ViewModel would live

**Comment by AndroidBob** [comment]
Build [544](https://10.10.124.134/job/AR-build-pipeline/544/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.3.0(7081543)(7081543)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.3.0(7081543)(7081543)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 9]
```diff
@@ -118,7 +119,8 @@ fun ConversationScreen(
     conversationBannerViewModel: ConversationBannerViewModel = hiltSavedStateViewModel(backNavArgs = backNavArgs),
     conversationCallViewModel: ConversationCallViewModel = hiltSavedStateViewModel(backNavArgs = backNavArgs),
     conversationMessagesViewModel: ConversationMessagesViewModel = hiltSavedStateViewModel(backNavArgs = backNavArgs),
-    messageComposerViewModel: MessageComposerViewModel = hiltSavedStateViewModel(backNavArgs = backNavArgs)
+    messageComposerViewModel: MessageComposerViewModel = hiltSavedStateViewModel(backNavArgs = backNavArgs),
+    compositeMessagesViewModel: CompositeMessagesViewModel = viewModel()
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]: Just curious, Shouldn't we use `hiltViewModel()` 🤔 ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt (hunk 3 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/CompositeMessagesViewModel.kt:32]: the idea here is to have this mutableStateMap to keep track of which buttons have been clicked.
the next PR will introduce type alias to make improve readability `Map<MessageId, ButtonID>`
this idea here is only one button per message is allowed to be clicked at a time.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt (hunk 4 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt:488]: My idea behind this is to handle it as a normal text message because it is just a text message with buttons.

**Comment by AndroidBob** [comment]
Build [554](https://10.10.124.134/job/AR-build-pipeline/554/) **failed**.

File: app/src/main/kotlin/com/wire/android/mapper/RegularMessageContentMapper.kt [hunk 10]
```diff
@@ -91,6 +92,34 @@ class RegularMessageMapper @Inject constructor(
...
+            UIMessageContent.Composite(
+                messageBody = text,
+                buttonList = content.buttonList.map {
+                    MessageButton(
+                        id = it.id,
+                        text = it.text,
+                        isSelected = it.isSelected
+                    )
+                },
+                deliveryStatus = mapRecipientsFailure(userList, message.deliveryStatus)
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/mapper/RegularMessageContentMapper.kt]: We can assume failure in general for composite messages in federated environments. Or in other words, we are required to show partial delivery failures only for: text messages, assets, and audio messages.

You can take a look at `Knock` for example or likes, no errors are displayed to the user.

**Review by yamilmedina — COMMENTED** [review]
Added a comment regarding `partial delivery` recipients mapper, per requirement we only consider, text messages and assets as partial deliverable, meaning only for those we show to whom we cannot deliver.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt [hunk 11]
```diff
@@ -212,6 +212,12 @@ sealed class UIMessageContent {
         override val deliveryStatus: DeliveryStatusContent = DeliveryStatusContent.CompleteDelivery
     ) : Regular(), PartialDeliverable
 
+    data class Composite(
+        val messageBody: MessageBody?,
+        val buttonList: List<MessageButton>,
+        override val deliveryStatus: DeliveryStatusContent = DeliveryStatusContent.CompleteDelivery
+    ) : Regular(), PartialDeliverable
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt]:
```suggestion
    ) : Regular()
```

No need to implement `PartialDeliverable` as this type is out of scope for partial deliverable.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt (hunk 8 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt:152]: i get you idea but this viewModel is supossed to be global for all messages so i will still need to pass lambda down down the path to message body 

tho i will play around with this idea in a following PR and try to to make a ViewModel for single message and see how it goes

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt (hunk 11 above)
Comment (MohamadJaara) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt]: my assumption is since composite messages contain text it should be handled as text
anyway there is no way in the app to send composite messages 👍🏼

**Review by yamilmedina — APPROVED** [review]
Thanks, so quick 😎

**Comment by AndroidBob** [comment]
Build [559](https://10.10.124.134/job/AR-build-pipeline/559/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.3.0(7088163)(7088163)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.3.0(7088163)(7088163)-staging-compat.apk)
